File: csharp/BullDB/Observability.cs
using System;
using System.Collections.Generic;

namespace BullDB
{
    public class ObservabilityEngine
    {
        private readonly Dictionary<string, long> _metrics = new Dictionary<string, long>();
        private readonly object _lock = new object();

        public static ObservabilityEngine Telemetry { get; } = new ObservabilityEngine();

        public void RecordQueryMetrics(string query, TimeSpan duration)
        {
            lock (_lock)
            {
                IncrementMetric("query_executions");
                Console.WriteLine($"[TELEMETRY] Query: {query}. Duration: {duration}");
            }
        }

        public void IncrementMetric(string name)
        {
            lock (_lock)
            {
                if (!_metrics.ContainsKey(name)) _metrics[name] = 0;
                _metrics[name]++;
            }
        }

        public long GetMetric(string name)
        {
            lock (_lock)
            {
                return _metrics.TryGetValue(name, out var val) ? val : 0;
            }
        }
    }

    public class TracingSpan : IDisposable
    {
        public string Name { get; }
        public DateTime StartTime { get; }

        public TracingSpan(string name)
        {
            Name = name;
            StartTime = DateTime.UtcNow;
        }

        public void Dispose()
        {
            var elapsed = DateTime.UtcNow - StartTime;
            Console.WriteLine($"[TRACING] Span {Name} finished in {elapsed}");
        }
    }
}

File: rust/bulldb/src/observability.rs
use std::collections::HashMap;
use std::sync::OnceLock;
use std::sync::Mutex;
use std::time::{Duration, Instant};

pub struct ObservabilityEngine {
    metrics: Mutex<HashMap<String, i64>>,
}

static TELEMETRY: OnceLock<ObservabilityEngine> = OnceLock::new();

fn get_telemetry() -> &'static ObservabilityEngine {
    TELEMETRY.get_or_init(|| ObservabilityEngine {
        metrics: Mutex::new(HashMap::new()),
    })
}

pub fn record_query_metrics(query: &str, duration: Duration) {
    let t = get_telemetry();
    let mut m = t.metrics.lock().unwrap();
    let count = m.entry("query_executions".to_string()).or_insert(0);
    *count += 1;
    println!("[TELEMETRY] Query: {}. Duration: {:?}", query, duration);
}

pub fn increment_metric(name: &str) {
    let t = get_telemetry();
    let mut m = t.metrics.lock().unwrap();
    let count = m.entry(name.to_string()).or_insert(0);
    *count += 1;
}

pub fn get_metric(name: &str) -> i64 {
    let t = get_telemetry();
    let m = t.metrics.lock().unwrap();
    *m.get(name).unwrap_or(&0)
}

pub struct TracingSpan {
    name: String,
    start_time: Instant,
}

impl TracingSpan {
    pub fn start(name: &str) -> Self {
        Self {
            name: name.to_string(),
            start_time: Instant::now(),
        }
    }

    pub fn finish(self) {
        println!("[TRACING] Span {} finished in {:?}", self.name, self.start_time.elapsed());
    }
}
